File: backend/firestore/test/ticket_deletion.test.mjs
/**
 * A2: eliminar un gasto.
 *
 * Dos fronteras y una evidencia. Quién puede borrar (su creador —que en este
 * modelo es el dueño de la sesión— y quien administra el GRUPO; una relación
 * no adquiere esa figura) y cuándo (nunca con la sesión cerrada). Y sobre
 * todo: el borrado y la evidencia de QUIÉN lo hizo viajan en el MISMO commit,
 * porque el trigger de borrado solo recibe el cambio neto y sin evidencia P6
 * atribuiría el hecho al dueño de la sesión — falsamente y para siempre.
 *
 * Ejecutar desde la raíz del repo:
 *   firebase emulators:exec --only firestore,storage --project demo-salda \
 *     "npm --prefix backend/firestore test"
 */
import { readFileSync } from 'node:fs';
import { after, before, beforeEach, describe, it } from 'node:test';
import {
  assertFails,
  assertSucceeds,
  initializeTestEnvironment,
} from '@firebase/rules-unit-testing';
import {
  Timestamp,
  deleteDoc,
  doc,
  getDoc,
  serverTimestamp,
  setDoc,
  updateDoc,
  writeBatch,
} from 'firebase/firestore';

let env;

const JEFA = 'uid-jefa'; // propietaria del grupo
const ADMIN = 'uid-admin'; // role: admin en el grupo
const ALBA = 'uid-alba'; // subió el ticket = dueña de la sesión
const JORGE = 'uid-jorge'; // miembro normal del grupo
const AJENA = 'uid-ajena'; // administra OTRO grupo
const PAREJA = 'uid-pareja'; // la otra mitad de la relación

const SG = 'sessions/sg1'; // sesión del GRUPO
const SR = 'sessions/sr1'; // sesión de la RELACIÓN

const db = (uid) =>
  env
    .authenticatedContext(uid, {
      email: `${uid}@salda.test`,
      email_verified: true,
      firebase: { sign_in_provider: 'password' },
    })
    .firestore();

/** El borrado REAL: evidencia + delete en un solo commit. */
const eliminar = (actor, { sid = 'sg1', tid = 't1', aid = 'a1', ...campos } = {}) => {
  const f = db(actor);
  const batch = writeBatch(f);
  batch.set(doc(f, `sessions/${sid}/ticketRemovals/${tid}`), {
    ticketId: tid,
    accountId: aid,
    merchantName: 'Familycash',
    grandTotal: 1596,
    removedBy: actor,
    removedAt: serverTimestamp(),
    schemaVersion: 1,
    ...campos,
  });
  batch.delete(doc(f, `sessions/${sid}/accounts/${aid}/tickets/${tid}`));
  return batch.commit();
};

before(async () => {
  env = await initializeTestEnvironment({
    projectId: 'demo-salda',
    firestore: {
      rules: readFileSync(
        new URL('../firestore.rules', import.meta.url),
        'utf8',
      ),
    },
  });
});
after(async () => env?.cleanup());

beforeEach(async () => {
  await env.clearFirestore();
  await env.withSecurityRulesDisabled(async (ctx) => {
    const f = ctx.firestore();
    for (const uid of [JEFA, ADMIN, ALBA, JORGE, AJENA, PAREJA]) {
      await setDoc(doc(f, `profiles/${uid}`), {
        displayName: uid, displayNameLower: uid, username: uid,
        createdAt: serverTimestamp(), updatedAt: serverTimestamp(),
        schemaVersion: 1,
      });
    }

    await setDoc(doc(f, 'spaces/gr1'), {
      name: 'Piso', ownerUid: JEFA, kind: 'group', status: 'active',
      createdAt: serverTimestamp(), schemaVersion: 2,
    });
    for (const uid of [JEFA, ADMIN, ALBA, JORGE]) {
      await setDoc(doc(f, `spaces/gr1/members/${uid}`), {
        uid, joinedAt: serverTimestamp(),
        ...(uid === ADMIN ? { role: 'admin' } : {}),
      });
    }
    // Otro grupo: su administradora no manda aquí.
    await setDoc(doc(f, 'spaces/gr2'), {
      name: 'Otro', ownerUid: AJENA, kind: 'group', status: 'active',
      createdAt: serverTimestamp(), schemaVersion: 2,
    });
    await setDoc(doc(f, `spaces/gr2/members/${AJENA}`), {
      uid: AJENA, joinedAt: serverTimestamp(),
    });
    // Relación: sin administradores, ni siquiera para su propietaria.
    await setDoc(doc(f, 'spaces/rel1'), {
      name: 'Pareja', ownerUid: JEFA, kind: 'relationship',
      relationshipUids: [ALBA, PAREJA].sort(), status: 'active',
      createdAt: serverTimestamp(), schemaVersion: 2,
    });
    for (const uid of [ALBA, PAREJA]) {
      await setDoc(doc(f, `spaces/rel1/members/${uid}`), {
        uid, joinedAt: serverTimestamp(),
      });
    }

    for (const [sid, spaceId] of [['sg1', 'gr1'], ['sr1', 'rel1']]) {
      await setDoc(doc(f, `sessions/${sid}`), {
        ownerUid: ALBA, kind: 'multi', name: 'Compra', status: 'open',
        splitModeDefault: 'byItem', shareCode: 'SECRET-CODE-16CHARS',
        currency: 'EUR', contextModelVersion: 1, spaceId,
        computeVersion: 0, totals: { grandTotal: 1596 }, balances: {},
      });
      await setDoc(doc(f, `sessions/${sid}/accounts/a1`), {
        name: 'Súper', order: 0,
      });
      await setDoc(doc(f, `sessions/${sid}/accounts/a1/tickets/t1`), {
        kind: 'scanned', grandTotal: 1596, paidByParticipantId: 'p1',
        merchant: { name: 'Familycash' }, spaceId, contextModelVersion: 1,
      });
      await setDoc(doc(f, `sessions/${sid}/accounts/a1/tickets/t1/lines/l1`), {
        name: 'Patatas', totalPrice: 1596, quantityMilli: 1000, order: 0,
      });
    }
  });
});

describe('A2: autoridad para eliminar un gasto', () => {
  it('el creador del ticket (dueño de la sesión) borra el suyo en un grupo',
    () => assertSucceeds(eliminar(ALBA)));

  it('el creador del ticket borra el suyo en una relación', () =>
    assertSucceeds(eliminar(ALBA, { sid: 'sr1' })));

  it('quien administra el grupo borra el ticket ajeno', () =>
    assertSucceeds(eliminar(ADMIN, { removedBy: ADMIN })));

  it('el propietario del grupo borra el ticket ajeno', () =>
    assertSucceeds(eliminar(JEFA, { removedBy: JEFA })));

  it('un miembro normal NO borra el ticket de otro', () =>
    assertFails(eliminar(JORGE, { removedBy: JORGE })));

  it('la contraparte de una RELACIÓN no borra el ticket de la otra', () =>
    assertFails(eliminar(PAREJA, { sid: 'sr1', removedBy: PAREJA })));

  it('la propietaria de la RELACIÓN tampoco: no hay administración', () =>
    assertFails(eliminar(JEFA, { sid: 'sr1', removedBy: JEFA })));

  it('quien administra OTRO grupo no borra aquí', () =>
    assertFails(eliminar(AJENA, { removedBy: AJENA })));

  it('con la sesión CERRADA nadie borra, ni su dueña', async () => {
    await env.withSecurityRulesDisabled((ctx) => updateDoc(
      doc(ctx.firestore(), SG), { status: 'closed' }));
    await assertFails(eliminar(ALBA));
    await assertFails(eliminar(ADMIN, { removedBy: ADMIN }));
  });
});

describe('A2: atomicidad y honestidad de la evidencia', () => {
  it('borrar SIN dejar evidencia: denegado', () =>
    assertFails(deleteDoc(doc(db(ALBA), `${SG}/accounts/a1/tickets/t1`))));

  it('dejar evidencia SIN borrar: denegado', () =>
    assertFails(setDoc(doc(db(ALBA), `${SG}/ticketRemovals/t1`), {
      ticketId: 't1', accountId: 'a1', merchantName: 'Familycash',
      grandTotal: 1596, removedBy: ALBA, removedAt: serverTimestamp(),
      schemaVersion: 1,
    })));

  it('el actor no se puede falsear', () =>
    assertFails(eliminar(ADMIN, { removedBy: ALBA })));

  it('la hora no se puede falsear', () =>
    assertFails(eliminar(ALBA, { removedAt: Timestamp.fromMillis(1) })));

  it('el importe tiene que ser el del ticket', () =>
    assertFails(eliminar(ALBA, { grandTotal: 100 })));

  it('el comercio tiene que ser el del ticket', () =>
    assertFails(eliminar(ALBA, { merchantName: 'Inventado' })));

  it('la evidencia no puede señalar otra cuenta', () =>
    assertFails(eliminar(ALBA, { accountId: 'a9' })));

  it('no hay evidencia de un ticket que no existe', () =>
    assertFails(setDoc(doc(db(ALBA), `${SG}/ticketRemovals/t9`), {
      ticketId: 't9', accountId: 'a1', merchantName: 'Fantasma',
      grandTotal: 0, removedBy: ALBA, removedAt: serverTimestamp(),
      schemaVersion: 1,
    })));

  it('la evidencia es INMUTABLE: ni update ni delete', async () => {
    await assertSucceeds(eliminar(ALBA));
    const ruta = `${SG}/ticketRemovals/t1`;
    await assertFails(updateDoc(doc(db(ALBA), ruta), { removedBy: JORGE }));
    await assertFails(deleteDoc(doc(db(ALBA), ruta)));
    await assertFails(deleteDoc(doc(db(JEFA), ruta)));
  });

  it('la lee quien podía ver el gasto, y nadie más', async () => {
    await assertSucceeds(eliminar(ALBA));
    const ruta = `${SG}/ticketRemovals/t1`;
    await assertSucceeds(getDoc(doc(db(ALBA), ruta))); // dueña de la sesión
    await assertSucceeds(getDoc(doc(db(JORGE), ruta))); // miembro del grupo
    await assertFails(getDoc(doc(db(AJENA), ruta))); // ajena al grupo
  });

  it('la evidencia NO abre el ticket: es auditoría, no papelera', async () => {
    // Se siembra un ticket que sobrevive y una evidencia de otro: leer la
    // evidencia nunca puede convertirse en leer contenido.
    await assertSucceeds(eliminar(ALBA));
    await assertFails(
      getDoc(doc(db(AJENA), `${SG}/accounts/a1/tickets/t1`)));
  });
});

File: apps/mobile/test/ticket_deletion_test.dart
import 'package:domain/domain.dart';
import 'package:fake_cloud_firestore/fake_cloud_firestore.dart';
import 'package:firebase_core/firebase_core.dart' show FirebaseException;
import 'package:flutter/material.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:salda_mobile/features/sessions/application/session_providers.dart';
import 'package:salda_mobile/features/sessions/data/firestore_session_repository.dart';
import 'package:salda_mobile/features/sessions/domain/session_models.dart';
import 'package:salda_mobile/features/sessions/presentation/ticket_detail_screen.dart';
import 'package:salda_mobile/features/sessions/presentation/ticket_navigation.dart';
import 'package:salda_mobile/l10n/generated/app_localizations.dart';

import 'fakes.dart';

/// A2: eliminar un gasto.
///
/// Lo que se fija aquí es la mitad de cliente: a quién se le OFRECE la acción
/// —las Rules son la autoridad, esto solo decide qué se enseña—, qué se le
/// advierte antes de un borrado irreversible, y que el commit que sale es el
/// atómico (evidencia + borrado) y no un borrado suelto que P6 no podría
/// atribuir.
void main() {
  const ticketPath = 'sessions/s1/accounts/a1/tickets/t1';
  const ticket = SessionTicket(
    id: 't1',
    path: ticketPath,
    merchantName: 'Súper',
    grandTotal: Money(2000),
    paidBy: 'p1',
    kind: 'manual',
    spaceId: 'g1',
    contextModelVersion: 1,
  );

  late FakeFirebaseFirestore firestore;

  setUp(() => firestore = FakeFirebaseFirestore());

  Future<void> sembrar({
    String status = 'open',
    String kind = 'group',
    String ownerUid = 'owner',
  }) async {
    await firestore.doc('spaces/g1').set({
      'name': 'Piso',
      'ownerUid': 'jefa',
      'kind': kind,
      'relationshipUids': kind == 'relationship' ? ['owner', 'pareja'] : [],
      'status': 'active',
      'schemaVersion': 2,
    });
    for (final uid in ['jefa', 'owner', 'admin', 'miembro', 'pareja']) {
      await firestore.doc('spaces/g1/members/$uid').set({
        'uid': uid,
        if (uid == 'admin') 'role': 'admin',
      });
    }
    await firestore.doc('sessions/s1').set({
      'ownerUid': ownerUid,
      'kind': 'single',
      'name': 'Compra',
      'status': status,
      'splitModeDefault': 'byItem',
      'shareCode': 'TEST-CODE-1234567890',
      'currency': 'EUR',
      'spaceId': 'g1',
      'contextModelVersion': 1,
    });
    await firestore.doc('sessions/s1/accounts/a1').set({'name': 'Súper'});
    await firestore.doc(ticketPath).set({
      'kind': 'manual',
      'grandTotal': 2000,
      'paidByParticipantId': 'p1',
      'merchant': {'name': 'Súper'},
      'spaceId': 'g1',
      'contextModelVersion': 1,
    });
  }

  /// Obligación del ticket y un pago humano ligado a ella.
  Future<void> deuda({
    required String estado,
    int amount = 1000,
    String id = 'pay1',
  }) async {
    await firestore.doc('economicEntries/e1').set({
      'spaceId': 'g1',
      'debtorUid': 'otra',
      'creditorUid': 'owner',
      'amount': 2000,
      'currency': 'EUR',
      'memberUids': ['owner', 'otra'],
      'sessionId': 's1',
      'accountId': 'a1',
      'ticketId': 't1',
      'ticketName': 'Súper',
      'schemaVersion': 1,
    });
    await firestore.doc('economicPayments/$id').set({
      'memberUids': ['owner', 'otra'],
      'payerUid': 'otra',
      'receiverUid': 'owner',
      'amount': amount,
      'currency': 'EUR',
      'status': estado,
      'source': 'user',
      'allocations': {'e1': amount},
      'sessionIds': ['s1'],
      'schemaVersion': 1,
    });
  }

  Future<ProviderContainer> pump(
    WidgetTester tester, {
    String uid = 'owner',
  }) async {
    tester.view.physicalSize = const Size(800, 2000);
    tester.view.devicePixelRatio = 1.0;
    addTearDown(tester.view.reset);
    final container = ProviderContainer(
      overrides: [
        ...loggedInOverrides(firestore: firestore, uid: uid),
        // El emulador de pruebas no aplica Rules, y en producción SOLO la
        // dueña de la sesión la lee (ahí vive el `shareCode`). Sin esto
        // cualquier uid parecería el creador del gasto y la pantalla ofrecía
        // borrar a quien las Rules van a rechazar.
        if (uid != 'owner')
          sessionDetailProvider('s1').overrideWith(
            (ref) => Stream<SessionDetail?>.error(
              FirebaseException(
                plugin: 'cloud_firestore',
                code: 'permission-denied',
              ),
            ),
          ),
      ],
    );
    addTearDown(container.dispose);
    await tester.pumpWidget(
      UncontrolledProviderScope(
        container: container,
        child: MaterialApp(
          localizationsDelegates: AppLocalizations.localizationsDelegates,
          supportedLocales: AppLocalizations.supportedLocales,
          home: const TicketDetailScreen(
            ticket: TicketRef(
              sessionId: 's1',
              payerName: 'Edgar',
              ticket: ticket,
            ),
          ),
        ),
      ),
    );
    await tester.pumpAndSettle();
    return container;
  }

  Future<void> cerrar(WidgetTester tester) async {
    await tester.pumpWidget(const SizedBox.shrink());
    await tester.pump(const Duration(seconds: 2));
  }

  final accionBorrar = find.byIcon(Icons.delete_outline);

  group('a quién se le ofrece eliminar', () {
    testWidgets('el creador del gasto (dueño de la sesión), sí', (
      tester,
    ) async {
      await sembrar();
      await pump(tester);
      expect(accionBorrar, findsOneWidget);
      await cerrar(tester);
    });

    testWidgets('quien administra el grupo, sobre el gasto ajeno', (
      tester,
    ) async {
      await sembrar();
      await pump(tester, uid: 'admin');
      expect(accionBorrar, findsOneWidget);
      await cerrar(tester);
    });

    testWidgets('el propietario del grupo, sobre el gasto ajeno', (
      tester,
    ) async {
      await sembrar();
      await pump(tester, uid: 'jefa');
      expect(accionBorrar, findsOneWidget);
      await cerrar(tester);
    });

    testWidgets('un miembro normal, NO', (tester) async {
      await sembrar();
      await pump(tester, uid: 'miembro');
      expect(accionBorrar, findsNothing);
      await cerrar(tester);
    });

    testWidgets('la contraparte de una RELACIÓN, NO', (tester) async {
      await sembrar(kind: 'relationship');
      await pump(tester, uid: 'pareja');
      expect(accionBorrar, findsNothing);
      await cerrar(tester);
    });

    testWidgets('con la sesión CERRADA, a nadie', (tester) async {
      await sembrar(status: 'closed');
      await pump(tester);
      expect(accionBorrar, findsNothing);
      await cerrar(tester);
    });
  });

  group('confirmación', () {
    testWidgets('sin pagos: confirmación destructiva normal', (tester) async {
      await sembrar();
      await pump(tester);
      await tester.tap(accionBorrar);
      await tester.pumpAndSettle();

      expect(find.text('¿Eliminar este gasto?'), findsOneWidget);
      expect(
        find.textContaining('los balances se recalcularán'),
        findsOneWidget,
      );
      // Nada de pagos: no se inventa una advertencia que no toca.
      expect(find.textContaining('pagos registrados'), findsNothing);
      expect(find.textContaining('pendientes de confirmación'), findsNothing);
      await cerrar(tester);
    });

    testWidgets('con pago confirmado: aviso reforzado con importe', (
      tester,
    ) async {
      await sembrar();
      await deuda(estado: 'confirmed');
      await pump(tester);
      await tester.tap(accionBorrar);
      await tester.pumpAndSettle();

      expect(find.text('Este gasto tiene pagos registrados'), findsOneWidget);
      expect(find.textContaining('1 pago confirmado de 10,00'), findsOneWidget);
      expect(find.textContaining('invertirse'), findsOneWidget);
      await cerrar(tester);
    });

    testWidgets('con declaración pendiente: se distingue del pago confirmado', (
      tester,
    ) async {
      await sembrar();
      await deuda(estado: 'pending', amount: 2000);
      await pump(tester);
      await tester.tap(accionBorrar);
      await tester.pumpAndSettle();

      // La declaración NO se cuenta como pago confirmado: son dos hechos
      // distintos y mezclarlos mentiría sobre el saldo.
      expect(find.text('Este gasto tiene pagos registrados'), findsNothing);
      expect(find.textContaining('1 pago declarado de 20,00'), findsOneWidget);
      await cerrar(tester);
    });

    testWidgets('cancelar no borra nada', (tester) async {
      await sembrar();
      await pump(tester);
      await tester.tap(accionBorrar);
      await tester.pumpAndSettle();
      await tester.tap(find.text('Cancelar'));
      await tester.pumpAndSettle();

      expect((await firestore.doc(ticketPath).get()).exists, isTrue);
      await cerrar(tester);
    });
  });

  group('el borrado', () {
    testWidgets('confirmar borra el ticket y deja la evidencia del actor', (
      tester,
    ) async {
      await sembrar();
      await pump(tester, uid: 'admin');
      await tester.tap(accionBorrar);
      await tester.pumpAndSettle();
      await tester.tap(find.text('Eliminar'));
      await tester.pumpAndSettle();

      expect((await firestore.doc(ticketPath).get()).exists, isFalse);
      final evidencia = await firestore
          .doc('sessions/s1/ticketRemovals/t1')
          .get();
      expect(evidencia.exists, isTrue);
      // El actor es quien borra, no el dueño de la sesión: es lo único que
      // impide que P6 atribuya el hecho a quien creó el gasto.
      expect(evidencia.data()!['removedBy'], 'admin');
      expect(evidencia.data()!['accountId'], 'a1');
      // El resumen se copia del ticket: las Rules lo comparan con el real.
      expect(evidencia.data()!['merchantName'], 'Súper');
      expect(evidencia.data()!['grandTotal'], 2000);
      expect(find.text('Gasto eliminado'), findsOneWidget);
      await cerrar(tester);
    });

    testWidgets('un fallo no puede parecer un éxito', (tester) async {
      await sembrar();
      final container = ProviderContainer(
        overrides: loggedInOverrides(
          firestore: firestore,
          sessionRepository: _RepositorioQueFalla(firestore),
        ),
      );
      addTearDown(container.dispose);
      tester.view.physicalSize = const Size(800, 2000);
      tester.view.devicePixelRatio = 1.0;
      addTearDown(tester.view.reset);
      await tester.pumpWidget(
        UncontrolledProviderScope(
          container: container,
          child: MaterialApp(
            localizationsDelegates: AppLocalizations.localizationsDelegates,
            supportedLocales: AppLocalizations.supportedLocales,
            home: const TicketDetailScreen(
              ticket: TicketRef(
                sessionId: 's1',
                payerName: 'Edgar',
                ticket: ticket,
              ),
            ),
          ),
        ),
      );
      await tester.pumpAndSettle();
      await tester.tap(accionBorrar);
      await tester.pumpAndSettle();
      await tester.tap(find.text('Eliminar'));
      await tester.pumpAndSettle();

      expect(find.textContaining('No se pudo eliminar'), findsOneWidget);
      expect(find.text('Gasto eliminado'), findsNothing);
      // La pantalla sigue donde estaba y el gasto también.
      expect(find.byType(TicketDetailScreen), findsOneWidget);
      expect((await firestore.doc(ticketPath).get()).exists, isTrue);
      await cerrar(tester);
    });

    testWidgets('borrar dos veces no es un error', (tester) async {
      await sembrar();
      final repo = FirestoreSessionRepository(
        firestore: firestore,
        uid: () => 'owner',
        shareCodeFactory: () => 'X',
      );
      await repo.deleteTicket(ticketPath);
      await repo.deleteTicket(ticketPath);
      expect((await firestore.doc(ticketPath).get()).exists, isFalse);
    });
  });

  group('un gasto eliminado ya no se abre', () {
    testWidgets(
      'quien no tiene acceso ve «ya no está disponible», no un error',
      (tester) async {
        // Un ex-miembro (A11d) pierde con el gasto su derecho histórico: ni
        // llega por él, ni puede listar las cuentas. Antes acababa en un error
        // de carga genérico que invitaba a reintentar para siempre.
        await sembrar();
        await firestore.doc(ticketPath).delete();
        final container = ProviderContainer(
          overrides: [
            ...loggedInOverrides(firestore: firestore, uid: 'ex-miembro'),
            accountsProvider('s1').overrideWith(
              (ref) => Stream<List<SessionAccount>>.error(
                FirebaseException(
                  plugin: 'cloud_firestore',
                  code: 'permission-denied',
                ),
              ),
            ),
          ],
        );
        addTearDown(container.dispose);
        tester.view.physicalSize = const Size(800, 2000);
        tester.view.devicePixelRatio = 1.0;
        addTearDown(tester.view.reset);
        await tester.pumpWidget(
          UncontrolledProviderScope(
            container: container,
            child: MaterialApp(
              localizationsDelegates: AppLocalizations.localizationsDelegates,
              supportedLocales: AppLocalizations.supportedLocales,
              home: const TicketRoute(sessionId: 's1', ticketId: 't1'),
            ),
          ),
        );
        await tester.pumpAndSettle();

        expect(find.text('Este gasto ya no está disponible'), findsOneWidget);
        await cerrar(tester);
      },
    );
  });
}

/// Repositorio que rechaza el borrado, como harían las Rules con una sesión
/// cerrada o sin autoridad.
class _RepositorioQueFalla extends FirestoreSessionRepository {
  _RepositorioQueFalla(FakeFirebaseFirestore firestore)
    : super(
        firestore: firestore,
        uid: () => 'owner',
        shareCodeFactory: () => 'X',
      );

  @override
  Future<void> deleteTicket(String ticketPath) => Future<void>.error(
    FirebaseException(plugin: 'cloud_firestore', code: 'permission-denied'),
  );
}
